src/main/java/net/snowflake/client/core/HttpClientSettingsKey.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2012-2021 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.client.core;
+
+import java.io.Serializable;
+import org.apache.http.HttpHost;
+
+/**
+ * This class defines all non-static parameters needed to create an HttpClient object. It is used as
+ * the key for the static hashmap of reusable http clients.
+ */
+public class HttpClientSettingsKey implements Serializable {
+
+  private OCSPMode ocspMode;
+  private boolean useProxy;
+  private String proxyHost = "";
+  private int proxyPort = 0;
+  private String nonProxyHosts = "";
+  private String proxyUser = "";
+  private String proxyPassword = "";
+
+  public HttpClientSettingsKey(
+      OCSPMode mode, String host, int port, String nonProxyHosts, String user, String password) {
+    this.useProxy = true;
+    this.ocspMode = mode != null ? mode : OCSPMode.FAIL_OPEN;
+    this.proxyHost = host.trim();
+    this.proxyPort = port;
+    this.nonProxyHosts = nonProxyHosts.trim();
+    this.proxyUser = user.trim();
+    this.proxyPassword = password.trim();
+  }
+
+  public HttpClientSettingsKey(OCSPMode mode) {
+    this.useProxy = false;
+    this.ocspMode = mode != null ? mode : OCSPMode.FAIL_OPEN;
+  }
+
+  @Override
+  public boolean equals(final Object O) {
+    if (O instanceof HttpClientSettingsKey) {
+      HttpClientSettingsKey comparisonKey = (HttpClientSettingsKey) O;
+      if (comparisonKey.ocspMode.getValue() == this.ocspMode.getValue()) {
+        if (!comparisonKey.useProxy) {
+          return true;
+        } else if (comparisonKey.proxyHost.equalsIgnoreCase(this.proxyHost)
+            && comparisonKey.proxyPort == this.proxyPort
+            && comparisonKey.proxyUser.equalsIgnoreCase(this.proxyUser)
+            && comparisonKey.proxyPassword.equalsIgnoreCase(this.proxyPassword)) {
+          // update nonProxyHost if changed
+          if (!this.nonProxyHosts.equalsIgnoreCase(comparisonKey.nonProxyHosts)) {
```

Comment:
The nonProxyHosts field is not included in the equality check because it's easy to have a different String value represent the same functionality (ex: `.baz.com|.foo.com` vs `.foo.com|.baz.com`) and we don't want to hashmap to balloon in size because of this risk and degrade performance.

Instead, the nonProxyHosts field is overwritten each time.